docs(aws): AWS EC2 provider instructions, diagram and code for manual AWS IAM set up.

aleon1220 opened the pull request:
Something happened in [PR1964](https://github.com/spinnaker/spinnaker.github.io/pull/1964) i am re-doing the changes and re-committing them.
Please help @dorbin @brian-armory

aleon1220 committed:
branch patch from PR1964

Something happened in [PR1964](https://github.com/spinnaker/spinnaker.github.io/pull/1964) i am readding the changes and re-committing them.
Please help @dorbin @brian-armory

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,22 +91,31 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
 
-## Option-3 : Use AWS Console UI (Manual Steps) 
+## Option-3 : Configure with AWS IAM Console
+### Create Roles
```
@aleon1220 is there a plan for adding content under these level-3 headings?

aleon1220 replied:
hello Dave. Yes there is a plan. I grabbed some screenshots but now i think is best to just write direct code there.

dorbin replied:
So you're going to add it later, in a separate PR?

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,22 +91,31 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
 
 ## Halyard Configurations
-After the AWS IAM user, roles, policies and trust relationship have been set up, the next step is to add the AWS configurations to Spinnaker via Halyard CLI:
+After the AWS IAM  structure (user, roles, policies and trust relationship) has been set up, the next step is to add the AWS configurations to Spinnaker via Halyard CLI:
```
```suggestion
After the AWS IAM  structure (user, roles, policies, and trust relationship) has been set up, the next step is to add the AWS configurations to Spinnaker via Halyard CLI:
```

dorbin replied:
Also: "user, roles, policies, and trust relationship"

Some of those are plural and some are singular. If it's just one user and one trust relationship, and there can be multiple roles and policies, then that's ok. Otherwise, make them all plural.

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,22 +91,31 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
-After the AWS IAM user, roles, policies and trust relationship have been set up, the next step is to add the AWS configurations to Spinnaker via Halyard CLI:
+After the AWS IAM  structure (user, roles, policies and trust relationship) has been set up, the next step is to add the AWS configurations to Spinnaker via Halyard CLI:
 
+The General steps are the following:
```
You can delete this whole line. The previous paragraph is sufficient to introduce the steps.

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -119,21 +134,25 @@ hal config provider aws account add $AWS_ACCOUNT_NAME \
 
-You can view the available configuration flags for AWS within the
-[Halyard reference](/reference/halyard/commands#hal-config-provider-aws-account-add).
+You can view the available configuration flags for Spinnaker AWS provider within the
```
```suggestion
You can view the available configuration flags for the Spinnaker AWS provider within the
```

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -11,13 +11,19 @@ sidebar:
+
+Spinnaker will use an [AWS IAM structure](https://aws.amazon.com/iam/) of users, roles, policies, and so on, to access AWS services and resources securely. There are 3 options to set up the AWS IAM structure
+
+1. AWS CloudFormation templates deployed with the CloudFormation Console
```
I still think it would better to hyperlink each of these to its respective section below.

aleon1220 replied:
i can do that in the next iteration. I dont really know how to do it. what do you think?

dorbin replied:
Well, the syntax for doing this is very simple, but the anchor text to use is automatically generated, and will be long and complex in this case because the headings have lots of words and numbers. If you're set up to preview changes, as describe in this repo's [readme](https://github.com/spinnaker/spinnaker.github.io#readme), you can go into the generated html and look at the anchor text.

I'll do a suggested change, just to show you how, and I'll take a guess at the anchor text, but you should confirm the text that's generated.

dorbin reviewed (commented):
@aleon1220 thanks for re-doing this!

I've left some suggestions. Sorry if some of them are repetitive.

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

great suggestion thanks.

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

grammar fix

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 committed:
Merge branch 'master' into patch-1

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -11,13 +11,19 @@ sidebar:
+
+Spinnaker will use an [AWS IAM structure](https://aws.amazon.com/iam/) of users, roles, policies, and so on, to access AWS services and resources securely. There are 3 options to set up the AWS IAM structure
+
+1. AWS CloudFormation templates deployed with the CloudFormation Console
```
```suggestion
1. [AWS CloudFormation templates deployed with the CloudFormation Console](#option-1-configure-with-aws-cloudformation-console)
```

dorbin replied:
@aleon1220 here's how to hyperlink it.

aleon1220 committed:
Merge branch 'master' into patch-1

Commit:
Improved wording and removed redundant line.

dorbin reviewed (commented):
@aleon1220 Oops, looks like I responded some days ago but didn't submit my review. Looks like just the one comment about linking.

Commit:
Adds instructions with an example
the instructions allow Spinnaker operators to
deploy to AWS Accounts develop and staging.
arranged the wording of the instructions.

Commit:
Adds the AWS provider diagram with extension .drawio
The diagram can be opened online with https://www.draw.io/

Commit:
Adds the image in .svg format for the example
The example shows an architecture with 3 AWS accounts
1 Spinnaker Managing where Spinnaker Lives
2 develop
3 staging

aleon1220 commented:
Hello, @dorbin 
 I have added enough detail to make the AWS EC2 provider work. I added an example all throughout the instructions to make it simple to understand. A diagram that i added in format .svg based on the .drawio diagram done by @anshrma back in 2018. These instructions work in AWS as this is what i have working. It was based on documentation and experience found throughout the community. Thanks to Armory for writing a wonderful guide on the aws provider. They have instructions for set up with halyard and operator.

Many of the new links are not working because they are not created but in general terms i checked this page after running it locally in my docker container dev set up.

![image](https://user-images.githubusercontent.com/2084646/90854466-1c425280-e3d1-11ea-8664-87e8ffc5cf15.png)

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 commented:
@dorbin  Hi Dave, I have added details for the EC2 provider. By using it with these instructions one can successfully deploy Instances to EC2. It is a big change because i think his document was not updated much. In summary there are 3 options to set up AWs IAM structure for Spinnaker.
1-AWS CF templates (this doesnt work)
2-AWS CLI to deploy the CF templates (this doesnt work)
3-Set up AWS IAM manually. These are the instructions i am contributing. Is been tested 4 times in my environments. It works but it requires many steps. I used an example to make it simple to understand and relate.

dorbin commented:
@brian-armory can you take a look at this when you get a chance? I'm not on AWS, so I can't test these instrux. Thanks!

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 committed:
Merge branch 'master' into patch-1

317brian commented:
Hi @aleon1220, I've scheduled time for myself to work on this during this sprint (this week and next). Sorry for the delay.

aleon1220 commented:
> Hi @aleon1220, I've scheduled time for myself to work on this during this sprint (this week and next). Sorry for the delay.

Thanks Brian. I am still preparing my AWS personal account to test this and record a short video. Other than that. Thanks for your review.

aleon1220 committed:
Fixes anchor link to new section AWS IAM user or roles

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -7,17 +7,25 @@ sidebar:
 {% include toc %}
 
-> :warning: These instructions are out-of-date and a new version is being
+> :warning: These instructions were updated to manually set up the AWS provider with [Option-3](#option-3--configure-with-aws-iam-console) on **2020-08-22**. <br>
```
```suggestion
> :warning: These instructions were updated, on **2020-08-22**, to manually set up the AWS provider with [Option-3](#option-3--configure-with-aws-iam-console). <br>
```

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,55 +93,234 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
 
-## Option-3 : Use AWS Console UI (Manual Steps) 
+## Option-3 : Configure with AWS IAM Console
+The Spinnaker **Managing** role AWS account Assumes the Spinnaker **Managed** role AWS account in a target AWS account via AWS IAM resources (policies, roles, users, trust relationship, etc). This allows Spinnaker to take control of the AWS cloud resources.
```
```suggestion
The Spinnaker **Managing** role AWS account assumes the Spinnaker **Managed** role AWS account in a target AWS account via AWS IAM resources (policies, roles, users, trust relationship, etc). This allows Spinnaker to control the AWS cloud resources.
```

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,55 +93,234 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+## Option-3 : Configure with AWS IAM Console
+The Spinnaker **Managing** role AWS account Assumes the Spinnaker **Managed** role AWS account in a target AWS account via AWS IAM resources (policies, roles, users, trust relationship, etc). This allows Spinnaker to take control of the AWS cloud resources.
+
+For the Example below the AWS Account **spinnakerManaging** assumes the **spinnakerManaged** role in the AWS accounts **develop** and **staging**. The account **spinnakerManaging** is where Spinnaker lives
```
```suggestion
For the Example below the AWS Account **spinnakerManaging** assumes the **spinnakerManaged** role in the AWS accounts **develop** and **staging**. The account **spinnakerManaging** is where Spinnaker lives.
```

dorbin replied:
"...where Spinnaker _lives_."

Is there a more descriptive term than "lives"? Is it the account Spinnaker _uses_ to access AWS resources?

aleon1220 replied:
@dorbin  Is the Account where Spinnaker is deployed and running.

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,55 +93,234 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+
+![Example AWS IAM structure for Spinnaker AWS Provider](/setup/install/providers/images/example-aws-provider.svg)
 
+Before you start create a simple table mapping the Account names and account IDs for your desired set up. An example table is shown below.
```
```suggestion
Before you start, create a simple table that maps the Account names to account IDs for your desired set up. An example table is shown below.
```

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,55 +93,234 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+2. Using AWS IAM Roles. Option 2 uses the IAM roles *ManagingRole* and *ManagedRoles*. This setting is applied on some environments that have extra security considerations. For Example the EC2 instance where Spinnaker is used to deploy AWS resources has the *ManagingRole* attached. The *ManagingRole* contains the AWS IAM policies required to manage AWS resources.
+
+### Create AWS IAM user
+- Access AWS IAM console
```
@aleon1220 can you make this an ordered list instead of unordered?

aleon1220 replied:
yes

dorbin commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -85,55 +93,234 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+1. Create **spinnakerManaged** in **develop ID=200000000002**
+2. Create **spinnakerManaged** in **staging ID=300000000003**
+
+- Access AWS IAM console
```
Here too, it's a procedure.

dorbin reviewed (commented):
Hi, @aleon1220 . A few more suggestions. Also, it looks like you've got a conflict in one of the files.

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

Great suggestion

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 commented:
> Hi, @aleon1220 . A few more suggestions. Also, it looks like you've got a conflict in one of the files.

sadly yes. This PR is been here for a while. Maybe someone updated the page. I dont understand. This AWS provider really needs some attention.

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

simple and concise. Thanks

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

grammar fix.

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 committed:
Update setup/install/providers/aws/aws-ec2.md

grammar fix

Co-authored-by: Dave Dorbin <ddorbin@google.com>

aleon1220 committed:
Merges parts from PR2020

aleon1220 committed:
Merge branch 'patch-1' of github.com:aleon1220/spinnaker.github.io into patch-1

317brian committed:
Merge branch 'master' into patch-1

317brian committed:
WIP edits

aleon1220 committed:
Merge branch 'master' into patch-1

317brian committed:
edits

317brian commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -102,47 +113,252 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+
+1. Select the AWS account where Spinnaker lives **spinnakerManaging**.
+2. Access AWS IAM console.
+3. List the AWS ARN of the **spinnakerManaged** accounts **develop** and **staging**. For example, the list below shows 3 AWS ARNs for different AWS accounts, each corresponding to an environment:
```
@aleon1220 this sentence says 3 arns, but the example only shows two. Is the sentence incorrect or is the example missing one of the arns?

aleon1220 replied:
I have added it now with a bulleted list of AWS accounts for more redundancy.

317brian commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -102,47 +113,252 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+
+1. Navigate to the AWS IAM console. 
+2. Create the role.
+3. Select **EC2**, and click **Next: Permissions**
```
Is it intentional that the `BaseIAMRole` have no permission policies attached?

aleon1220 replied:
Great Spotting. I don't give permissions to that role because is a default one. If something in cloud Driver fails and it requires the use of the default role then errors will be generated.
Perhaps it should have basic permissions to interact with AWS EC2 AMIs or some default policy but I don't think it should have Administration powers of any kind.

- What do you think? @brian-armory 

Actually the BaseIAM role only has a default permissions policy that allows assumeRole
```
  BaseIAMRole:
    Type: AWS::IAM::Role
    Properties:
      RoleName: BaseIAMRole
      AssumeRolePolicyDocument:
        Statement:
          - Action:
              - sts:AssumeRole
            Effect: Allow
            Principal:
              Service:
                - ec2.amazonaws.com
        Version: "2012-10-17"
      Path: /
```

317brian replied:
Definitely not admin powers since it should function as a sane default, and erroring out on something obvious before a person gets too far makes sense to me.

317brian commented on setup/install/providers/aws/aws-ec2.md:
````diff
@@ -102,47 +113,252 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+
+   ```bash
+   # Adds spinnakerManaged for the regions us-east-1 and us-west-2
+   export AWS_ACCOUNT_NAME=""
````
Two comments:
* I think this is the correct syntax for setting a variable `export <variable-name> = "<some-name>"`. You had `$<variable-name>`. Let me know if that method actually works. I'm used to the `export` method.
* Is it intentional that the AWS_ACCOUNT_NAME is blank for spinnakerManaging (arn 100000000001)?

aleon1220 replied:
I have fixed those issues and added the example ARN

317brian commented on setup/install/providers/aws/aws-ec2.md:
````diff
@@ -102,47 +113,252 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+   The following examples add the ****spinnakerManaged** and **develop** accounts from the previous examples. Repeat the command for every account you want to add.
+
+   ```bash
+   # Adds spinnakerManaged for the regions us-east-1 and us-west-2
````
```suggestion
# Adds spinnakerManaging for the regions us-east-1 and us-west-2

317brian commented on setup/install/providers/aws/aws-ec2.md:
````diff
@@ -102,47 +113,252 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
-    --account-id ${ACCOUNT_ID} \
-    --assume-role role/spinnakerManaged
-```
+   The following examples add the ****spinnakerManaged** and **develop** accounts from the previous examples. Repeat the command for every account you want to add.
````
```suggestion
   The following examples add the ****spinnakerManaging** and **develop** accounts from the previous examples. Repeat the command for every account you want to add.

317brian committed:
Update setup/install/providers/aws/aws-ec2.md

317brian committed:
Update setup/install/providers/aws/aws-ec2.md

317brian committed:
fix anchor link

317brian committed:
Merge branch 'master' into patch-1

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 committed:
minor fixes and improved the halyard example to enable the AWS provider

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 committed:
Merge branch 'master' into patch-1

aleon1220 commented:
@brian-armory Hi Brian, Thanks for all the help. I have improved the doc based on your suggestions. I have been keeping this branch updated. Would you like to do a review?

317brian reviewed (approved):
LGTM. Thanks for all your hardwork on this @aleon1220. I pushed a commit to fix some indenting on lines 214-16 for the nested list.

317brian committed:
fix indent

aleon1220 reviewed (commented):
Thank you so much!

dorbin commented:
Thanks for all the work, Andres! And all the re-doing of the work. 😃

avram commented on setup/install/providers/aws/aws-ec2.md:
```diff
@@ -102,47 +113,257 @@ aws cloudformation deploy --stack-name spinnaker-managed-infrastructure-setup --
+2. Switch to the AWS Account you want to create the roll for.
+3. Go to **Roles > Create Role**.
+4. Select EC2. You can change this later, because we want to specify an explicit consumer of this role in a later stage.
+5. For permissions, search for "PowerUserAccess" and select this policy. This gives the role permission to access AWS services.
```
PowerUserAccess is broader access than Spinnaker really needs; there used to be a very specific policy document doc that had what Spinnaker needs; we should refer to that here. I think that Armory at least has an up-to-date, narrower policy to refer to.

aleon1220 replied:
what access do you recommend? 
the policy is AWS managed it should be ok...

avram replied:
The PowerUserAccess policy is extremely broad. Unfortunately it looks like the Armory docs also now recommend PowerUserAccess.

This is the policy I'm using now, which at least limits access to the core services Spinnaker needs, although it would need to be broadened to account for CloudFormation and Lambda and potentially other small revisions. This or a similar policy used to be present in the Spinnaker docs (or perhaps the Armory docs).
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1486065689000",
            "Effect": "Allow",
            "Action": [
                "autoscaling:CreateAutoScalingGroup",
                "autoscaling:CreateLaunchConfiguration",
                "autoscaling:CreateOrUpdateTags",
                "autoscaling:DeleteAutoScalingGroup",
                "autoscaling:DeleteLaunchConfiguration",
                "autoscaling:DeletePolicy",
                "autoscaling:DeleteScheduledAction",
                "autoscaling:DeleteTags",
                "autoscaling:DescribeAutoScalingGroups",
                "autoscaling:DescribeLaunchConfigurations",
                "autoscaling:DescribeLoadBalancers",
                "autoscaling:DescribePolicies",
                "autoscaling:DescribeScalingActivities",
                "autoscaling:DescribeScheduledActions",
                "autoscaling:DetachInstances",
                "autoscaling:DisableMetricsCollection",
                "autoscaling:EnableMetricsCollection",
                "autoscaling:PutLifecycleHook",
                "autoscaling:PutNotificationConfiguration",
                "autoscaling:PutScalingPolicy",
                "autoscaling:PutScheduledUpdateGroupAction",
                "autoscaling:ResumeProcesses",
                "autoscaling:SuspendProcesses",
                "autoscaling:TerminateInstanceInAutoScalingGroup",
                "autoscaling:UpdateAutoScalingGroup",
                "cloudwatch:DeleteAlarms",
                "cloudwatch:DescribeAlarms",
                "cloudwatch:GetMetricStatistics",
                "cloudwatch:ListMetrics",
                "cloudwatch:PutMetricAlarm",
                "ec2:AttachClassicLinkVpc",
                "ec2:AuthorizeSecurityGroupIngress",
                "ec2:CreateNetworkInterface",
                "ec2:CreateSecurityGroup",
                "ec2:CreateTags",
                "ec2:DeleteSecurityGroup",
                "ec2:DeleteTags",
                "ec2:DescribeAccountAttributes",
                "ec2:DescribeAddresses",
                "ec2:DescribeAvailabilityZones",
                "ec2:DescribeClassicLinkInstances",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "ec2:DescribeKeyPairs",
                "ec2:DescribeRegions",
                "ec2:DescribeReservedInstances",
                "ec2:DescribeReservedInstancesOfferings",
                "ec2:DescribeSecurityGroups",
                "ec2:DescribeSpotPriceHistory",
                "ec2:DescribeSubnets",
                "ec2:DescribeTags",
                "ec2:DescribeVpcClassicLink",
                "ec2:DescribeVpcs",
                "ec2:GetConsoleOutput",
                "ec2:ModifyImageAttribute",
                "ec2:ModifyInstanceAttribute",
                "ec2:RebootInstances",
                "ec2:RevokeSecurityGroupIngress",
                "ec2:TerminateInstances",
                "elasticloadbalancing:ApplySecurityGroupsToLoadBalancer",
                "elasticloadbalancing:ConfigureHealthCheck",
                "elasticloadbalancing:CreateListener",
                "elasticloadbalancing:CreateLoadBalancer",
                "elasticloadbalancing:CreateLoadBalancerListeners",
                "elasticloadbalancing:CreateLoadBalancerPolicy",
                "elasticloadbalancing:CreateRule",
                "elasticloadbalancing:CreateTargetGroup",
                "elasticloadbalancing:DeleteListener",
                "elasticloadbalancing:DeleteLoadBalancer",
                "elasticloadbalancing:DeleteLoadBalancerListeners",
                "elasticloadbalancing:DeleteRule",
                "elasticloadbalancing:DeleteTargetGroup",
                "elasticloadbalancing:DeregisterInstancesFromLoadBalancer",
                "elasticloadbalancing:DeregisterTargets",
                "elasticloadbalancing:DescribeInstanceHealth",
                "elasticloadbalancing:DescribeListeners",
                "elasticloadbalancing:DescribeLoadBalancerAttributes",
                "elasticloadbalancing:DescribeLoadBalancerPolicies",
                "elasticloadbalancing:DescribeLoadBalancers",
                "elasticloadbalancing:DescribeRules",
                "elasticloadbalancing:DescribeTargetGroupAttributes",
                "elasticloadbalancing:DescribeTargetGroups",
                "elasticloadbalancing:DescribeTargetHealth",
                "elasticloadbalancing:ModifyListener",
                "elasticloadbalancing:ModifyLoadBalancerAttributes",
                "elasticloadbalancing:ModifyTargetGroup",
                "elasticloadbalancing:ModifyTargetGroupAttributes",
                "elasticloadbalancing:RegisterInstancesWithLoadBalancer",
                "elasticloadbalancing:RegisterTargets",
                "elasticloadbalancing:SetLoadBalancerPoliciesOfListener",
                "elasticloadbalancing:SetSecurityGroups",
                "iam:ListServerCertificates",
                "iam:PassRole"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

aleon1220 replied:
A customer managed policy is cumbersome for an organization. Is an interesting concept to try to reduce risk but is just too much to handle.

avram replied:
With cloudformation stacksets and AWS organizations it is quite feasible to have a customer managed policy.

The power user policy is an irresponsible one and it reflects poorly on the product to recommend it.

aleon1220 replied:
@avram I understand. If that's the case I highly encourage you to improve the Cloud formation templates that use the AWS IAM policies at https://github.com/spinnaker/spinnaker.github.io/tree/master/downloads/aws
